### ISSUE TITLE
Support related resource responses in DELETE requests

### ISSUE THREAD

**Issue by cmeeren** [json-api/json-api#1567]
Imagine a "parent" and "child" resource, like an order with order lines.

```
order
  totalPrice: attribute
  orderlines: relationship, to-many orderline

orderline
  order: relationship, to-one order
```

When you `DELETE /orderlines/{id}`, the parent order's `totalPrice` changes. With the current specification, the client must `GET /orders/{id}` after the `DELETE` in order to get the updated order data. It would be better if it was possible to include the parent resource after deleting the child resource, e.g. if `DELETE /orderlines/{id}` could in some way respond with the updated order.

(For the record, #1492 also deals with including resources in responses where currently unsupported, but the issues are distinct.)

**Comment by cmeeren** [comment]
Is this something that could be considered for a future version of the spec (perhaps even 1.1), or are there fundamental problems with it?

**Comment by jelhan** [comment]
_I mixed up #1492 and this one. I didn't even noticed after @cmeeren mentioned it the first time. Please excuse any confusion caused by that._

> Is this something that could be considered for a future version of the spec (perhaps even 1.1), or are there fundamental problems with it?

~The proposal to use `.` as a keyword to identify target resource in a relationship update seems reasonable to me. See https://github.com/json-api/json-api/issues/1492#issuecomment-662275093 for more details on that proposal. A pull request would be helpful to identify any potential problems we haven't noticed so far.~

~I think it is unlikely that it gets into v1.1. Main priority is to ship v1.1 finally to unlock experiments using extensions and profiles as far as I understood it. But beside limited time from editors and contributors nothing prevents us from shipping a v1.2 soon afterwards.~

**Comment by cmeeren** [comment]
> The proposal to use `.` as a keyword to identify target resource if a relationship update seems reasonable to me.

Did you mean to post this in #1492?

**Comment by jelhan** [comment]
_I mixed up https://github.com/json-api/json-api/issues/1492 and this one. I didn't even noticed after @cmeeren mentioned it the first time. Please excuse any confusion caused by that._

> > The proposal to use `.` as a keyword to identify target resource if a relationship update seems reasonable to me.
> 
> Did you mean to post this in #1492?

~If I didn't missed something, both #1492 and this issue would be solved by introducing a keyword to identify the target resource of a relationship upgrade in include path, isn't it?~

~To include the target resource of a relationship upgrade, which is mainly discussed in #1492, `?include=.` would work. To include a related resource, a relationship path based on the target resource could be used. In this case it would be `?include=.order`.~

**Comment by cmeeren** [comment]
> To include a related resource, a relationship path based on the target resource could be used. In this case it would be `?include=.order`.

Why not just `include=order`? Here, it makes no sense to include the target resource, which is the resource you just deleted. Therefore, starting a path with `.` (with the assumption that this would include the current resource, which has just been deleted) would be invalid.

To be clear:

* `PATCH /orderlines/{id}?include=order` returns the modified orderline as primary data and the order as included data
* `PATCH /orderlines/{id}/relationships/whatever?include=.` returns the relationship identifiers as primary data and the modified orderline as included data (and `include=.order` would return both the orderline and its order as included data)
* `DELETE /orderlines/{id}?include=order` (my suggestion) returns empty primary data (since the orderline has been deleted) and its (before-deletion) parent order as included data
* `DELETE /orderlines/{id}?include=.order` (your suggestion), if keeping the semantics from the relationship update, would indicate that the (deleted) orderline as well as its order should be returned as included data, which is invalid since the orderline doesn't exist anymore and therefore can't be returned.

Did that make sense? Please let me know if anything is unclear.

**Comment by jelhan** [comment]
You are right. I confused #1492 and this one. Will move the comments over. :+1:

**Comment by cmeeren** [comment]
For the clarity of future readers, are you able to mark these comments as hidden?

**Comment by cmeeren** [comment]
Hi, I still haven't heard back regarding this. Could it be considered for inclusion in the spec? I think this is useful enough that it warrants inclusion in the base spec and not an extension. (Is an extension even allowed to do this?)

**Comment by jelhan** [comment]
> Could it be considered for inclusion in the spec? I think this is useful enough that it warrants inclusion in the base spec

It is considered for inclusion in base specification. But I'm not sure yet how to do this.

I think informing a client about side-effects caused by a mutation needs to be addressed in general. Not only for deletion of resources.

A server can use a compound document to inform a client about side-effects using a compound document when creating and updating a resource. But also this patterns has noticeable limitations:

1. It only works if there is a relationship between primary resource(s) and the resources updated as a side-effect. This is often the case but may require including interim resources.
2. A server can only inform a client about current state of resources by including them in the response. The server cannot explicitly inform the client that resources changed. Client would need doing a diff from old state it was aware of and the state in the response to detect if there was a change. This requires including all changed attributes of the updated resource(s). Often a client does not need the new representation right away. It might be enough to know that a locally cached representation of the resource is outdated and that the resource must be refreshed when it is needed again.
3. Informing a client about side-effects using a compound document does not work well together with `included` query parameter. A server must not include any other related resources than those specified by `included` query parameter if the query parameter is set. This prevents the server from informing the client about side-effects in that case.
4. It does not work for relationship updates using a relationship URL as discussed in #1492.

All in all I think we should explore better ways to inform a client about side-effects. We should do that exploration first in an extension before adding something to the base specification.

**Comment by jelhan** [comment]
> I think we should explore better ways to inform a client about side-effects. We should do that exploration first in an extension before adding something to the base specification.

Please see https://github.com/json-api/json-api/issues/1492#issuecomment-1207826202 for early thoughts how such an extension may look like.
